File: sdk/typescript/src/namespaces/agent-selection.ts
/**
 * Agent Selection Namespace API
 *
 * Provides methods for agent team selection operations:
 * - Plugin discovery and configuration
 * - Agent scoring for specific tasks
 * - Team selection with role assignment
 * - Selection history tracking
 */

// =============================================================================
// Plugin Types
// =============================================================================

/**
 * Selection plugin information.
 */
export interface SelectionPlugin {
  name: string;
  type: 'scorer' | 'team_selector' | 'role_assigner';
  description: string;
  version: string;
  enabled: boolean;
  config_schema: Record<string, unknown>;
  default_config?: Record<string, unknown>;
}

/**
 * Scorer plugin details.
 */
export interface ScorerPlugin extends SelectionPlugin {
  type: 'scorer';
  metrics: string[];
  weight_range: { min: number; max: number };
}

/**
 * Team selector plugin details.
 */
export interface TeamSelectorPlugin extends SelectionPlugin {
  type: 'team_selector';
  strategies: string[];
  supports_constraints: boolean;
}

/**
 * Role assigner plugin details.
 */
export interface RoleAssignerPlugin extends SelectionPlugin {
  type: 'role_assigner';
  roles: string[];
  assignment_strategies: string[];
}

/**
 * Default plugin configuration.
 */
export interface DefaultPluginConfig {
  scorer: string;
  team_selector: string;
  role_assigner: string;
  scorer_weights: Record<string, number>;
}

/**
 * List plugins response.
 */
export interface ListPluginsResponse {
  plugins: SelectionPlugin[];
}

// =============================================================================
// Agent Scoring Types
// =============================================================================

/**
 * Agent score request (matches Python SDK).
 */
export interface ScoreAgentsRequest {
  /** List of agent identifiers to score */
  agents: string[];
  /** Task context or description for scoring */
  context?: string;
  /** Scoring dimensions (e.g., ["accuracy", "speed", "cost"]) */
  dimensions?: string[];
  /** Specific scorer plugin to use */
  scorer?: string;
  /** Custom weights for each dimension */
  weights?: Record<string, number>;
  /** Return only top K agents */
  top_k?: number;
}

/**
 * Agent score result.
 */
export interface AgentScore {
  agent_id: string;
  agent_name: string;
  overall_score: number;
  dimension_scores: Record<string, number>;
  confidence: number;
  reasoning: string;
}

/**
 * Score agents response.
 */
export interface ScoreAgentsResponse {
  scores: AgentScore[];
  scorer_used: string;
  scoring_time_ms: number;
}

/**
 * Best agent request.
 */
export interface GetBestAgentRequest {
  /** List of candidate agent identifiers */
  pool: string[];
  /** Type of task (e.g., "code_review", "analysis", "creative") */
  task_type: string;
  /** Additional context for selection */
  context?: string;
}

/**
 * Best agent response.
 */
export interface GetBestAgentResponse {
  agent_id: string;
  agent_name: string;
  score: number;
  reasoning: string;
  task_type: string;
}

// =============================================================================
// Team Selection Types
// =============================================================================

/**
 * Team selection request (matches Python SDK).
 */
export interface SelectTeamRequest {
  /** List of candidate agent identifiers */
  pool: string[];
  /** Dict describing task needs (domain, complexity, etc.) */
  task_requirements?: Record<string, unknown>;
  /** Exact team size */
  team_size?: number;
  /** Additional selection constraints */
  constraints?: Record<string, unknown>;
  /** Minimum team size */
  min_team_size?: number;
  /** Maximum team size */
  max_team_size?: number;
  /** Roles that must be filled */
  required_roles?: string[];
  /** Agents to exclude from selection */
  excluded_agents?: string[];
  /** Weight for team diversity (0.0-1.0) */
  diversity_weight?: number;
  /** Specific team selector plugin to use */
  selector?: string;
  /** Specific role assigner plugin to use */
  role_assigner?: string;
}

/**
 * Team member with assigned role.
 */
export interface TeamMember {
  agent_id: string;
  agent_name: string;
  role: string;
  score: number;
  strengths: string[];
  selection_reasoning: string;
}

/**
 * Team selection response.
 */
export interface SelectTeamResponse {
  team: TeamMember[];
  team_score: number;
  diversity_score: number;
  coverage_score: number;
  selector_used: string;
  role_assigner_used: string;
  selection_time_ms: number;
  alternatives?: TeamMember[][];
}

// =============================================================================
// Role Assignment Types
// =============================================================================

/**
 * Role assignment request.
 */
export interface AssignRolesRequest {
  /** List of agent identifiers to assign roles to */
  members: string[];
  /** List of roles to assign */
  roles: string[];
  /** Context for role assignment decisions */
  task_context?: string;
  /** Specific role assigner plugin to use */
  assigner?: string;
}

/**
 * Role assignment entry.
 */
export interface RoleAssignment {
  agent_id: string;
  agent_name: string;
  role: string;
  confidence: number;
  reasoning: string;
}

/**
 * Role assignment response.
 */
export interface AssignRolesResponse {
  assignments: RoleAssignment[];
  assigner_used: string;
  assignment_time_ms: number;
}

// =============================================================================
// History Types
// =============================================================================

/**
 * Selection history entry.
 */
export interface SelectionHistoryEntry {
  id: string;
  timestamp: string;
  selection_type: 'score' | 'team' | 'role';
  request: Record<string, unknown>;
  result: Record<string, unknown>;
  duration_ms: number;
}

/**
 * Selection history response.
 */
export interface SelectionHistoryResponse {
  history: SelectionHistoryEntry[];
  total: number;
  has_more: boolean;
}

/**
 * Interface for the internal client methods used by AgentSelectionAPI.
 */
export interface AgentSelectionClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: unknown }
  ): Promise<T>;
}

/**
 * Agent Selection API namespace.
 *
 * Provides methods for agent team selection and scoring:
 * - Listing available selection plugins
 * - Scoring agents for specific tasks
 * - Selecting optimal teams with role assignment
 * - Managing plugin configurations
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // List available plugins
 * const { plugins } = await client.agentSelection.listPlugins();
 *
 * // Score agents for a task
 * const scores = await client.agentSelection.scoreAgents({
 *   task: 'Code review for security vulnerabilities',
 *   top_k: 5,
 * });
 *
 * // Select a team
 * const team = await client.agentSelection.selectTeam({
 *   task: 'Design system architecture',
 *   team_size: 4,
 *   required_roles: ['architect', 'security_expert', 'domain_expert'],
 * });
 * ```
 */
export class AgentSelectionAPI {
  constructor(private client: AgentSelectionClientInterface) {}

  /**
   * List all available selection plugins.
   */
  async listPlugins(): Promise<{ plugins: SelectionPlugin[] }> {
    return this.client.request<{ plugins: SelectionPlugin[] }>('GET', '/api/v1/selection/plugins');
  }

  /**
   * Get default plugin configuration.
   */
  async getDefaults(): Promise<DefaultPluginConfig> {
    return this.client.request<DefaultPluginConfig>('GET', '/api/v1/selection/defaults');
  }

  /**
   * Get details for a specific scorer plugin.
   */
  async getScorer(name: string): Promise<ScorerPlugin> {
    return this.client.request<ScorerPlugin>('GET', `/api/v1/selection/scorers/${name}`);
  }

  /**
   * Get details for a specific team selector plugin.
   */
  async getTeamSelector(name: string): Promise<TeamSelectorPlugin> {
    return this.client.request<TeamSelectorPlugin>('GET', `/api/v1/selection/team-selectors/${name}`);
  }

  /**
   * Get details for a specific role assigner plugin.
   */
  async getRoleAssigner(name: string): Promise<RoleAssignerPlugin> {
    return this.client.request<RoleAssignerPlugin>('GET', `/api/v1/selection/role-assigners/${name}`);
  }

  /**
   * Score agents for a specific task.
   */
  async scoreAgents(request: ScoreAgentsRequest): Promise<ScoreAgentsResponse> {
    return this.client.request<ScoreAgentsResponse>('POST', '/api/v1/selection/score', {
      json: request,
    });
  }

  /**
   * Select an optimal team for a task.
   */
  async selectTeam(request: SelectTeamRequest): Promise<SelectTeamResponse> {
    return this.client.request<SelectTeamResponse>('POST', '/api/v1/selection/team', {
      json: request,
    });
  }

  /**
   * Convenience method to get the best agent for a task.
   */
  async getBestAgent(task: string, context?: string): Promise<AgentScore | null> {
    const response = await this.scoreAgents({ task, context, top_k: 1 });
    return response.scores[0] || null;
  }

  /**
   * Convenience method to select a team with alternatives.
   */
  async selectTeamWithAlternatives(
    request: SelectTeamRequest,
    alternativeCount: number = 2
  ): Promise<{
    primary: SelectTeamResponse;
    alternatives: SelectTeamResponse[];
  }> {
    // Select primary team
    const primary = await this.selectTeam(request);

    // Select alternatives by excluding primary team members
    const alternatives: SelectTeamResponse[] = [];
    let excludedAgents = request.excluded_agents || [];

    for (let i = 0; i < alternativeCount; i++) {
      // Add previous team members to exclusion list
      const previousTeamIds = primary.team.map(m => m.agent_id);
      excludedAgents = [...new Set([...excludedAgents, ...previousTeamIds])];

      try {
        const altTeam = await this.selectTeam({
          ...request,
          excluded_agents: excludedAgents,
        });
        alternatives.push(altTeam);

        // Add this team's members to exclusion for next iteration
        excludedAgents = [...excludedAgents, ...altTeam.team.map(m => m.agent_id)];
      } catch {
        // Not enough agents for more alternatives
        break;
      }
    }

    return { primary, alternatives };
  }

  /**
   * List all scorer plugins.
   */
  async listScorers(): Promise<{ scorers: ScorerPlugin[] }> {
    const { plugins } = await this.listPlugins();
    return {
      scorers: plugins.filter((p): p is ScorerPlugin => p.type === 'scorer'),
    };
  }

  /**
   * List all team selector plugins.
   */
  async listTeamSelectors(): Promise<{ selectors: TeamSelectorPlugin[] }> {
    const { plugins } = await this.listPlugins();
    return {
      selectors: plugins.filter((p): p is TeamSelectorPlugin => p.type === 'team_selector'),
    };
  }

  /**
   * List all role assigner plugins.
   */
  async listRoleAssigners(): Promise<{ assigners: RoleAssignerPlugin[] }> {
    const { plugins } = await this.listPlugins();
    return {
      assigners: plugins.filter((p): p is RoleAssignerPlugin => p.type === 'role_assigner'),
    };
  }
}

export default AgentSelectionAPI;

File: sdk/python/aragora_sdk/namespaces/agent_selection.py
"""
Agent Selection Namespace API

Provides methods for agent team selection operations:
- Plugin discovery and configuration
- Agent scoring for specific tasks
- Team selection with role assignment
- Selection history tracking
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class AgentSelectionAPI:
    """
    Synchronous Agent Selection API.

    Provides methods for agent team selection and scoring:
    - List available selection plugins
    - Get default plugin configurations
    - Score agents for specific tasks
    - Select optimal teams with role assignment
    - Track selection history

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> plugins = client.agent_selection.list_plugins()
        >>> scores = client.agent_selection.score_agents(
        ...     agents=["claude", "gpt-4", "gemini"],
        ...     context="security code review",
        ...     dimensions=["accuracy", "speed", "cost"],
        ... )
        >>> team = client.agent_selection.select_team(
        ...     pool=["claude", "gpt-4", "gemini", "mistral"],
        ...     task_requirements={"domain": "security", "complexity": "high"},
        ...     team_size=3,
        ... )
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Plugin Discovery
    # ===========================================================================

    def list_plugins(self) -> dict[str, Any]:
        """
        List all available selection plugins.

        Returns plugins for scoring, team selection, and role assignment.

        Returns:
            Dict with plugins array containing plugin metadata including
            name, type, description, version, enabled status, and config schema
        """
        return self._client.request("GET", "/api/v1/agent-selection/plugins")

    def get_defaults(self) -> dict[str, Any]:
        """
        Get default plugin configuration.

        Returns the default scorer, team selector, role assigner,
        and scorer weight settings.

        Returns:
            Dict with default plugin names and weight configurations
        """
        return self._client.request("GET", "/api/v1/agent-selection/defaults")

    # ===========================================================================
    # Agent Scoring
    # ===========================================================================

    def score_agents(
        self,
        agents: list[str],
        context: str | None = None,
        dimensions: list[str] | None = None,
        scorer: str | None = None,
        weights: dict[str, float] | None = None,
        top_k: int | None = None,
    ) -> dict[str, Any]:
        """
        Score agents for a specific task or context.

        Evaluates agents across multiple dimensions (e.g., accuracy, speed, cost)
        and returns ranked scores with reasoning.

        Args:
            agents: List of agent identifiers to score
            context: Task context or description for scoring
            dimensions: Scoring dimensions (e.g., ["accuracy", "speed", "cost"])
            scorer: Specific scorer plugin to use (optional)
            weights: Custom weights for each dimension (optional)
            top_k: Return only top K agents (optional)

        Returns:
            Dict with scores array containing agent scores, dimension breakdowns,
            confidence levels, and reasoning for each agent
        """
        data: dict[str, Any] = {"agents": agents}
        if context is not None:
            data["context"] = context
        if dimensions is not None:
            data["dimensions"] = dimensions
        if scorer is not None:
            data["scorer"] = scorer
        if weights is not None:
            data["weights"] = weights
        if top_k is not None:
            data["top_k"] = top_k
        return self._client.request("POST", "/api/v1/agent-selection/score", json=data)

    def get_best_agent(
        self,
        pool: list[str],
        task_type: str,
        context: str | None = None,
    ) -> dict[str, Any]:
        """
        Get the best agent for a specific task from a pool.

        Convenience method that returns the single best-suited agent
        for the given task type and context.

        Args:
            pool: List of candidate agent identifiers
            task_type: Type of task (e.g., "code_review", "analysis", "creative")
            context: Additional context for selection (optional)

        Returns:
            Dict with the best agent details including score and reasoning
        """
        data: dict[str, Any] = {
            "pool": pool,
            "task_type": task_type,
        }
        if context is not None:
            data["context"] = context
        return self._client.request("POST", "/api/v1/agent-selection/best", json=data)

    # ===========================================================================
    # Team Selection
    # ===========================================================================

    def select_team(
        self,
        pool: list[str],
        task_requirements: dict[str, Any] | None = None,
        team_size: int | None = None,
        constraints: dict[str, Any] | None = None,
        min_team_size: int | None = None,
        max_team_size: int | None = None,
        required_roles: list[str] | None = None,
        excluded_agents: list[str] | None = None,
        diversity_weight: float | None = None,
        selector: str | None = None,
        role_assigner: str | None = None,
    ) -> dict[str, Any]:
        """
        Select an optimal team of agents for a task.

        Analyzes the agent pool and selects a team optimized for
        the given task requirements, with automatic role assignment.

        Args:
            pool: List of candidate agent identifiers
            task_requirements: Dict describing task needs (domain, complexity, etc.)
            team_size: Exact team size (optional, use min/max for range)
            constraints: Additional selection constraints (optional)
            min_team_size: Minimum team size (optional)
            max_team_size: Maximum team size (optional)
            required_roles: Roles that must be filled (optional)
            excluded_agents: Agents to exclude from selection (optional)
            diversity_weight: Weight for team diversity (0.0-1.0, optional)
            selector: Specific team selector plugin to use (optional)
            role_assigner: Specific role assigner plugin to use (optional)

        Returns:
            Dict with team array containing selected members with roles,
            team score, diversity score, coverage score, and selection metadata
        """
        data: dict[str, Any] = {"pool": pool}
        if task_requirements is not None:
            data["task_requirements"] = task_requirements
        if team_size is not None:
            data["team_size"] = team_size
        if constraints is not None:
            data["constraints"] = constraints
        if min_team_size is not None:
            data["min_team_size"] = min_team_size
        if max_team_size is not None:
            data["max_team_size"] = max_team_size
        if required_roles is not None:
            data["required_roles"] = required_roles
        if excluded_agents is not None:
            data["excluded_agents"] = excluded_agents
        if diversity_weight is not None:
            data["diversity_weight"] = diversity_weight
        if selector is not None:
            data["selector"] = selector
        if role_assigner is not None:
            data["role_assigner"] = role_assigner
        return self._client.request("POST", "/api/v1/agent-selection/select-team", json=data)

    def assign_roles(
        self,
        members: list[str],
        roles: list[str],
        task_context: str | None = None,
        assigner: str | None = None,
    ) -> dict[str, Any]:
        """
        Assign roles to a set of team members.

        Given a list of agents and available roles, determines the
        optimal role assignment based on agent capabilities.

        Args:
            members: List of agent identifiers to assign roles to
            roles: List of roles to assign
            task_context: Context for role assignment decisions (optional)
            assigner: Specific role assigner plugin to use (optional)

        Returns:
            Dict with assignments mapping agents to roles, with reasoning
        """
        data: dict[str, Any] = {
            "members": members,
            "roles": roles,
        }
        if task_context is not None:
            data["task_context"] = task_context
        if assigner is not None:
            data["assigner"] = assigner
        return self._client.request("POST", "/api/v1/agent-selection/assign-roles", json=data)

    # ===========================================================================
    # History
    # ===========================================================================

    def get_selection_history(
        self,
        limit: int | None = None,
        since: str | None = None,
    ) -> dict[str, Any]:
        """
        Get agent selection history.

        Returns past selection decisions for analysis and auditing.

        Args:
            limit: Maximum number of history entries to return (optional)
            since: ISO 8601 timestamp to filter history from (optional)

        Returns:
            Dict with history array containing past selection decisions
        """
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if since is not None:
            params["since"] = since
        return self._client.request("GET", "/api/v1/agent-selection/history", params=params)

    def team_selection(self, **kwargs: Any) -> dict[str, Any]:
        """
        Get team selection overview.

        GET /api/v1/team-selection

        Returns:
            Dict with team selection data
        """
        return self._client.request("GET", "/api/v1/team-selection", params=kwargs or None)


class AsyncAgentSelectionAPI:
    """
    Asynchronous Agent Selection API.

    Provides async methods for agent team selection and scoring:
    - List available selection plugins
    - Get default plugin configurations
    - Score agents for specific tasks
    - Select optimal teams with role assignment
    - Track selection history

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     plugins = await client.agent_selection.list_plugins()
        ...     scores = await client.agent_selection.score_agents(
        ...         agents=["claude", "gpt-4", "gemini"],
        ...         context="security code review",
        ...         dimensions=["accuracy", "speed", "cost"],
        ...     )
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # Plugin Discovery
    # ===========================================================================

    async def list_plugins(self) -> dict[str, Any]:
        """List all available selection plugins."""
        return await self._client.request("GET", "/api/v1/agent-selection/plugins")

    async def get_defaults(self) -> dict[str, Any]:
        """Get default plugin configuration."""
        return await self._client.request("GET", "/api/v1/agent-selection/defaults")

    # ===========================================================================
    # Agent Scoring
    # ===========================================================================

    async def score_agents(
        self,
        agents: list[str],
        context: str | None = None,
        dimensions: list[str] | None = None,
        scorer: str | None = None,
        weights: dict[str, float] | None = None,
        top_k: int | None = None,
    ) -> dict[str, Any]:
        """Score agents for a specific task or context."""
        data: dict[str, Any] = {"agents": agents}
        if context is not None:
            data["context"] = context
        if dimensions is not None:
            data["dimensions"] = dimensions
        if scorer is not None:
            data["scorer"] = scorer
        if weights is not None:
            data["weights"] = weights
        if top_k is not None:
            data["top_k"] = top_k
        return await self._client.request("POST", "/api/v1/agent-selection/score", json=data)

    async def get_best_agent(
        self,
        pool: list[str],
        task_type: str,
        context: str | None = None,
    ) -> dict[str, Any]:
        """Get the best agent for a specific task from a pool."""
        data: dict[str, Any] = {
            "pool": pool,
            "task_type": task_type,
        }
        if context is not None:
            data["context"] = context
        return await self._client.request("POST", "/api/v1/agent-selection/best", json=data)

    # ===========================================================================
    # Team Selection
    # ===========================================================================

    async def select_team(
        self,
        pool: list[str],
        task_requirements: dict[str, Any] | None = None,
        team_size: int | None = None,
        constraints: dict[str, Any] | None = None,
        min_team_size: int | None = None,
        max_team_size: int | None = None,
        required_roles: list[str] | None = None,
        excluded_agents: list[str] | None = None,
        diversity_weight: float | None = None,
        selector: str | None = None,
        role_assigner: str | None = None,
    ) -> dict[str, Any]:
        """Select an optimal team of agents for a task."""
        data: dict[str, Any] = {"pool": pool}
        if task_requirements is not None:
            data["task_requirements"] = task_requirements
        if team_size is not None:
            data["team_size"] = team_size
        if constraints is not None:
            data["constraints"] = constraints
        if min_team_size is not None:
            data["min_team_size"] = min_team_size
        if max_team_size is not None:
            data["max_team_size"] = max_team_size
        if required_roles is not None:
            data["required_roles"] = required_roles
        if excluded_agents is not None:
            data["excluded_agents"] = excluded_agents
        if diversity_weight is not None:
            data["diversity_weight"] = diversity_weight
        if selector is not None:
            data["selector"] = selector
        if role_assigner is not None:
            data["role_assigner"] = role_assigner
        return await self._client.request("POST", "/api/v1/agent-selection/select-team", json=data)

    async def assign_roles(
        self,
        members: list[str],
        roles: list[str],
        task_context: str | None = None,
        assigner: str | None = None,
    ) -> dict[str, Any]:
        """Assign roles to a set of team members."""
        data: dict[str, Any] = {
            "members": members,
            "roles": roles,
        }
        if task_context is not None:
            data["task_context"] = task_context
        if assigner is not None:
            data["assigner"] = assigner
        return await self._client.request("POST", "/api/v1/agent-selection/assign-roles", json=data)

    # ===========================================================================
    # History
    # ===========================================================================

    async def get_selection_history(
        self,
        limit: int | None = None,
        since: str | None = None,
    ) -> dict[str, Any]:
        """Get agent selection history."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if since is not None:
            params["since"] = since
        return await self._client.request("GET", "/api/v1/agent-selection/history", params=params)

    async def team_selection(self, **kwargs: Any) -> dict[str, Any]:
        """Get team selection overview. GET /api/v1/team-selection"""
        return await self._client.request("GET", "/api/v1/team-selection", params=kwargs or None)
